### PR TITLE
[spirv] Migration to Spiregg v2

### PR DESCRIPTION
**Pull request by ehsannas**
There are still some cleanups left, but this a working version.

### CODE REVIEW

File: tools/clang/include/clang/SPIRV/SpirvInstruction.h [hunk 1]
```diff
@@ -1682,6 +1720,47 @@ class SpirvVectorShuffle : public SpirvInstruction {
...
+
+  SpirvInstruction *getStructure() const { return structure; }
+  uint32_t getArrayMember() const { return arrayMember; }
+
+private:
+  SpirvInstruction *structure;
+  uint32_t arrayMember;
+};
+
+class SpirvLineInfo : public SpirvInstruction {
```
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/SpirvInstruction.h]: Do we want to keep an in-memory representation for debug line info? Better to leave debug info attached to whatever entity they are annotating and have a pass write out the binaries for them directly. Probably add a TODO to clean up later?

File: tools/clang/lib/SPIRV/CapabilityVisitor.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+#include "clang/SPIRV/SpirvContext.h"
+#include "clang/SPIRV/SpirvVisitor.h"
+
+namespace clang {
+namespace spirv {
+
+class SpirvBuilder;
+
+class CapabilityVisitor : public Visitor {
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/CapabilityVisitor.h]: THIS IS NICE! :)

File: tools/clang/lib/SPIRV/EmitVisitor.cpp [hunk 3]
```diff
@@ -1083,6 +1471,13 @@ uint32_t EmitTypeHandler::emitType(const SpirvType *type,
       curTypeInst.push_back(paramTypeId);
     finalizeTypeInstruction();
   }
+  // Hybrid Types
+  // Note: The type lowering pass should lower all types to SpirvTypes.
+  // Therefore, if we find a hybrid type when going through the emitting pass,
+  // that is clearly a bug.
+  else if (const auto *hybridType = dyn_cast<HybridType>(type)) {
+    assert(false && "found hybrid type when emitting SPIR-V");
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitVisitor.cpp]: Use llvm_unreachable

File: tools/clang/lib/SPIRV/LiteralTypeVisitor.cpp [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+    tryToUpdateInstLitType(inst->getOperand1(), resultType);
+    // The shitf amount (arg2) cannot be a 64-bit type for a 32-bit base!
+    tryToUpdateInstLitType(inst->getOperand2(), resultType);
+    return true;
+  }
+
+  // The following operations have a boolean return type, so we cannot deduce
+  // anything about the operand type from the result type. However, the two
+  // operands in these operations must have the same bitwidth.
+  if (op == spv::Op::OpIEqual || op == spv::Op::OpINotEqual ||
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/LiteralTypeVisitor.cpp]: A free standing function with a `switch` statement would be better.

File: tools/clang/lib/SPIRV/LiteralTypeVisitor.h [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+#include "clang/SPIRV/SpirvContext.h"
+#include "clang/SPIRV/SpirvVisitor.h"
+
+namespace clang {
+namespace spirv {
+
+class SpirvBuilder;
+
+class LiteralTypeVisitor : public Visitor {
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/LiteralTypeVisitor.h]: AWESOME CLEANUP! :)

File: tools/clang/lib/SPIRV/LowerTypeVisitor.cpp [hunk 6]
```diff
@@ -8,27 +8,228 @@
...
+  }
+  return true;
+}
+
+bool LowerTypeVisitor::visitInstruction(SpirvInstruction *instr) {
+  const QualType astType = instr->getAstResultType();
+  const SpirvType *hybridType = instr->getResultType();
+
+  // Lower QualType to SpirvType
+  if (astType != QualType({})) {
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/LowerTypeVisitor.cpp]: Can just do `!astType`? Also other places.

File: tools/clang/lib/SPIRV/LowerTypeVisitor.cpp [hunk 7]
```diff
@@ -92,12 +293,15 @@ const SpirvType *LowerTypeVisitor::lowerType(QualType type,
...
+        // LowerTypeVisitor is invoked. We should error out if we encounter a
+        // literal type.
         case BuiltinType::LitInt:
+          // emitError("found literal int type when lowering types", srcLoc);
+          return spvContext.getUIntType(64);
         case BuiltinType::LitFloat: {
-          // TODO: analyze adjacent instructions for type hints
-          emitError("TODO: literal int/float", srcLoc);
-          return spvContext.getVoidType();
+          // emitError("found literal float type when lowering types", srcLoc);
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/LowerTypeVisitor.cpp]: Remove commented out code

File: tools/clang/lib/SPIRV/SpirvEmitter.cpp [hunk 8]
```diff
@@ -1,4 +1,4 @@
-//===------- SPIRVEmitter.h - SPIR-V Binary Code Emitter --------*- C++ -*-===//
+//===------- SpirvEmitter.h - SPIR-V Binary Code Emitter --------*- C++ -*-===//
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SpirvEmitter.cpp]: .cpp

File: tools/clang/lib/SPIRV/SpirvEmitter.cpp [hunk 9]
```diff
@@ -2184,47 +2025,20 @@ SpirvEvalInfo SPIRVEmitter::processCall(const CallExpr *callExpr) {
...
+  return retVal;
 }
 
-SpirvEvalInfo SPIRVEmitter::doCastExpr(const CastExpr *expr) {
+SpirvInstruction *SpirvEmitter::doCastExpr(const CastExpr *expr) {
   const Expr *subExpr = expr->getSubExpr();
   const QualType subExprType = subExpr->getType();
   const QualType toType = expr->getType();
 
-  // Unfortunately the front-end fails to deduce some types in certain cases.
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SpirvEmitter.cpp]: So nice to see these stuff gone!

File: tools/clang/lib/SPIRV/SpirvType.cpp [hunk 10]
```diff
@@ -27,45 +27,232 @@ bool ScalarType::classof(const SpirvType *t) {
...
+  case spv::Dim::Buffer:
+    dimStr = "buffer.";
+    break;
+  case spv::Dim::SubpassData:
+    dimStr = "subpass.";
+    break;
+  default:
+    break;
+  }
+  std::string name =
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SpirvType.cpp]: Use `std::ostringstream`

**Comment by ehsannas** [comment]
Thanks for your review! I'm sorry this PR was so big. I will merge this and address the comments in follow-up PRs.

File: tools/clang/lib/SPIRV/LowerTypeVisitor.cpp (hunk 6 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/LowerTypeVisitor.cpp]: apparently can't do this.

**Pull request merged**